### ISSUE TITLE
Guide data download failures, isMasterBackendConnected : error, connecting to backend

### ISSUE THREAD

**Issue by billmeek** [MythTV-Clients/MythTV-Android-Frontend#117]
Connecting to this user's backend via the internet worked well in the
evening, however this AM, I'm getting repeated errors when trying
to upload his program guide.

If the Away screen is exited and reentered, the attempts to download
the program guide are restarted and again fail after a few hours are
collected. Yet, going to other screens e.g. Upcoming and Recording
Rules data is downloaded OK.

LogCat: http://paste.ubuntu.com/1658452/ line 1201

**Comment by dmfrey** [comment]
Are have you been able to successfully and repeatedly reconnect to the users backend?

I am not sure if we can account for unreliable internet connections. Should the app have done something differently?

**Comment by billmeek** [comment]
Yes, I have no trouble connecting to his backend. And after this failure, I
can go to another screen and get other downloads.

Just to be clean (re-reading my original comment,) when I said "after a
few hours are collected" I was referring to hours of Guide data, not that
I had to wait hours for it to start syncing again.

I've also used my browser and done things like Myth/GetHostName as
fast as I can press ctrl-r e.g. refresh and had no failures.

I suspect this is a result of TCP RSTs that I've discussed in the past.
But without being able to capture protocol, I can't prove it.

I would suggest that the test for backend availability might be enhanced
to make more than one check. If all is well, there would be no impact on
data transfer. If GetHostName failed, maybe it could be retried. Or, could
calls to that function be passed with an argument(s) for how long to try 
and/or how many attempts to make? Just tossing out ideas.

**Comment by billmeek** [comment]
Testing this now (e.g. line 43 and 112)

```
http://paste.ubuntu.com/1675778/
```

Initial results are that after the 1st failure, only 1 more attempt is
made, and it succeeds.

**Comment by dmfrey** [comment]
that looks good.

One questions...do you think the count down in isMasterBackendConnected is
needed, since it calls isNetworkConnected first, which could already be
doing the 3 calls.  Then if isMasterBackendConnected fails and tries again,
it could call isNetworkConnected, which could fail 3 more times.

The number of network calls could grow considerably, since this gets
continually called as calls iterate (ie. downloading each hour of the day
in program guide)

On Mon, Feb 18, 2013 at 1:07 AM, billmeek notifications@github.com wrote:

> Testing this now (e.g. line 43 and 112)
> 
> http://paste.ubuntu.com/1675778/
> 
> Initial results are that after the 1st failure, only 1 more attempt is
> made, and it succeeds.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/MythTV-Clients/MythTV-Android-Frontend/issues/117#issuecomment-13707736.

## 

My Google Profile http://www.google.com/profiles/dmfrey

**Comment by billmeek** [comment]
Short answer: yes, the counter is required (it counts up now to make
more sensible Log() messages.)

I'm not sure if you were suggesting that the isMasterBackendConnected
logic be moved inside of isNetworkConnected, so I didn't change it.

I moved isNetworkConnected before the  while() and retested with a full
download (all 288 hours.) The failures happened in the early attempts,
and there was never more than one retry. 

When these transient failures happen, it's due to a failed connection with
the backend host. So, the GetHostName has to be repeated.

What I observed with downloading the program guide was that on the first
(e.g after three attempts) failure, it stopped and notified the user (with a
Toast.)

When I reentered the Home/Away screen, downloading picked up where it left off,
a nice touch, by the way.

I did retest on my own backend. I let the guide start downloading, and then
stopped the backend. Three exceptions fired and the attempt counter reached
three and the download gracefully stopped.

The root cause of the backend connection failures is still unknown, but this
seem to be a reasonable band-aid.

If you're OK with the above, I'll commit it.
